src/json-validation.ts

```diff
@@ -0,0 +1,208 @@
+  let m;
+  if ((m = error.message.match(/at position (\d+)/))) {
+    return Math.min(+m[1], doc.length);
+  }
+  if ((m = error.message.match(/at line (\d+) column (\d+)/))) {
+    return Math.min(doc.line(+m[1]).from + +m[2] - 1, doc.length);
+  }
+  return 0;
+}
+
+const errorMessage = (error: any) => {
+  let addendum;
+  switch (error.keyword) {
+    case "additionalProperties":
+      addendum = error.params.additionalProperty;
+      break;
+    case "enum":
+      addendum = error.params.allowedValues.join(", ");
+      break;
+  }
+  return [error.instancePath, error.message, addendum]
+    .filter((_) => _)
+    .join(" ");
+};
+
+export function getRangeForJSONErrors(
+  view: EditorView,
+  validate?: ValidateFunction
+): Diagnostic[] {
+  if(!validate) return []
+  // const validate = ajv.compile(schema);
+  // let parsed: ParsedJSON;
+
+  // const valid = validate(JSON.parse(view.state.doc.toString()));
+  // if (valid) {
+  //   return [];
+  // }
+
+  // first see if the json can parse
+  let json: any;
+  const text = view.state.doc.toString();
+
+  // ignore blank texts
+  if (!text) return [];
+
+  try {
+    json = JsonMap.parse(text);
+  } catch (error) {
+    if (error instanceof SyntaxError) {
+      // parse the error message and report the line numbers
+      const pos = getErrorPosition(error as SyntaxError, view.state.doc);
+
+      return [
+        {
+          from: pos,
+          to: pos,
+          severity: "error",
+          message: error.message
+        },
+      ];
+    }
+  }
+  let valid = false
+  try {
+   valid = validate(json.data);
+  }
+  catch {}
+
+  if (valid || !validate.errors) return [];
+  return validate.errors.reduce((acc: Diagnostic[], error) => {
+    let pointer = json.pointers[error.instancePath];
+
+    if (pointer) {
+      // console.log('pointer', pointer, error);
+      acc.push({
+        from: pointer.value.pos,
+        to: pointer.valueEnd.pos,
+        message: errorMessage(error),
+        severity: "error",
+      } as Diagnostic);
+    }
+
+    return acc;
+  }, []);
+  // const document = syntaxTree(view.state);
+
+  // const errorPathNames = validate.errors?.map((error) => {
+  //   if(error.instancePath) return error.instancePath.slice(1).replace(/\//g, '.')
+  //   if(error.keyword === 'additionalProperties') {
+  //     return error.params.additionalProperty
+  //   }
+  //   if(error.keyword === 'required') {
+  //     return error.params.missingProperty
+  //   }
+  // })
+
+  // const locations = {}
+
+  // document.iterate({
+  //   from: 0,
+  //   to: view.state.doc.length,
+  //   enter: (type, from, to) => {
+  //     console.log(type.type.name)
+  //     let key = ''
+  //     if(type.type.name === 'Object') {
+  //     }
+  //     if(type.type.name === 'PropertyName') {
+  //       const original = view.state.doc.sliceString(
+  //         type.from,
+  //         type.to
+  //       );
+  //       console.log(original)
```

Comment:
here i could print individual property names, but building the entire dot notated path of the object property was not going to be easy